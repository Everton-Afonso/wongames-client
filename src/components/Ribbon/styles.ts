'use client'

import styled, { css } from 'styled-components'

import { RibbonProps } from '.'

const wrapperModifiers = {
  primary: () => css`
    background-color: var(--primary);

    &::before {
      border-left-color: var(--primary);
      border-top-color: var(--primary);
      filter: brightness(50%);
    }
  `,

  secondary: () => css`
    background-color: var(--secondary);

    &::before {
      border-left-color: var(--secondary);
      border-top-color: var(--secondary);
      filter: brightness(50%);
    }
  `,

  small: () => css`
    font-size: var(--xsmall);
    height: 24px;
    padding: 0 16px;
    right: -13px;

    &::before {
      top: 24px;
      border-top-width: 6px;
      border-right-width: 13px;
    }
  `,

  normal: () => css`
    font-size: var(--small);
    height: 33px;
    padding: 0 16px;
    right: -14px;

    &::before {
      top: 33px;
      border-top-width: 8px;
      border-right-width: 14px;
    }
  `,

  large: () => css`
    font-size: var(--large);
    height: 40px;
    padding: 0 16px;
    right: -22px;

    &::before {
      top: 40px;
      border-top-width: 10px;
      border-right-width: 22px;
    }
  `
}

export const WrapperRibbon = styled.div<Omit<RibbonProps, 'children'>>`
  ${({ color, sizes }) => css`
    position: absolute;
    top: var(--xsmall-spacings);
    display: flex;
    align-items: center;
    font-weight: var(--bold);
    color: var(--white);

    &::before {
      content: '';
      position: absolute;
      right: 0;
      border-style: solid;
      border-left-width: 0rem;
      border-right-color: transparent;
      border-bottom-color: transparent;
      border-bottom-width: 1rem;
    }

    ${!!color && wrapperModifiers[color]}
    ${!!sizes && wrapperModifiers[sizes]}
  `}
`
