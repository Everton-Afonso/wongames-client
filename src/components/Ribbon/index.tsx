import * as S from './styles'

export type RibbonColors = 'primary' | 'secondary'
export type RibbonSizes = 'small' | 'normal' | 'large'

export type RibbonProps = {
  children: React.ReactNode
  color?: RibbonColors
  sizes?: RibbonSizes
}

const Ribbon = ({
  children,
  color = 'primary',
  sizes = 'normal'
}: RibbonProps) => {
  return (
    <S.WrapperRibbon color={color} sizes={sizes}>
      {children}
    </S.WrapperRibbon>
  )
}

export default Ribbon
