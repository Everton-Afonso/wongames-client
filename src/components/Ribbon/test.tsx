import { render, screen } from '../../utils/tests/helpers'

import Ribbon from '.'

describe('<Ribbon />', () => {
  it('should render the text conrrectly', () => {
    const { container } = render(<Ribbon>Best Seller</Ribbon>)

    expect(screen.getByText(/best seller/i)).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render with the primary color', () => {
    render(<Ribbon>Best Seller</Ribbon>)

    expect(screen.getByText(/best seller/i)).toHaveStyle({
      backgroundColor: '#F231A5'
    })
  })

  it('should render with the secondary color', () => {
    render(<Ribbon color="secondary">Best Seller</Ribbon>)

    expect(screen.getByText(/best seller/i)).toHaveStyle({
      backgroundColor: '#3CD3C1'
    })
  })

  it('should render with the small size', () => {
    render(<Ribbon sizes="small">Best Seller</Ribbon>)

    expect(screen.getByText(/best seller/i)).toHaveStyle({
      height: '24px',
      fontSize: '0.75rem'
    })
  })

  it('should render with the normal size as default', () => {
    render(<Ribbon>Best Seller</Ribbon>)

    expect(screen.getByText(/best seller/i)).toHaveStyle({
      height: '33px',
      fontSize: '0.875rem'
    })
  })

  it('should render with the large size', () => {
    render(<Ribbon sizes="large">Best Seller</Ribbon>)

    expect(screen.getByText(/best seller/i)).toHaveStyle({
      height: '40px',
      fontSize: '1.125rem'
    })
  })
})
