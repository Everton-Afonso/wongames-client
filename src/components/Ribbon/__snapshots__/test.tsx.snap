// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Ribbon /> should render the text conrrectly 1`] = `
.c0 {
  z-index: 1;
  position: absolute;
  top: 1rem;
  display: flex;
  align-items: center;
  font-weight: 600;
  color: #FAFAFA;
  background-color: #F231A5;
  font-size: 0.875rem;
  height: 33px;
  padding: 0 16px;
  right: -14px;
}

.c0::before {
  content: '';
  position: absolute;
  right: 0;
  border-style: solid;
  border-left-width: 0rem;
  border-right-color: transparent;
  border-bottom-color: transparent;
  border-bottom-width: 1rem;
}

.c0::before {
  border-left-color: #F231A5;
  border-top-color: #F231A5;
  filter: brightness(50%);
}

.c0::before {
  top: 33px;
  border-top-width: 8px;
  border-right-width: 14px;
}

<div
  class="c0"
  color="primary"
  sizes="normal"
>
  Best Seller
</div>
`;
